tests/kademlia/testputval.nim

```diff
@@ -0,0 +1,147 @@
+{.used.}
+import chronicles
+import strformat
+import sequtils
+import std/[times]
+# import std/enumerate
+import chronos
+import ../../libp2p/[switch, builders]
+import ../../libp2p/protocols/kademlia
+import ../../libp2p/protocols/kademlia/routingtable
+import ../../libp2p/protocols/kademlia/keys
+import ../../libp2p/protocols/kademlia/dhttypes
+import unittest2
+import ../utils/async_tests
+import std/tables
+
+proc createSwitch(): Switch =
+  SwitchBuilder
+  .new()
+  .withRng(newRng())
+  .withAddresses(@[MultiAddress.init("/ip4/0.0.0.0/tcp/0").tryGet()])
+  .withTcpTransport()
+  .withMplex()
+  .withNoise()
+  .build()
+
+proc countBucketEntries(buckets: seq[Bucket], key: Key): uint32 =
+  var res: uint32 = 0
+  for b in buckets:
+    for ent in b.peers:
+      if ent.nodeId == key:
+        res += 1
+  return res
+
+type PermissiveValidator = ref object of EntryValidator
+
+method validate(self: PermissiveValidator, cand: EntryCandidate): bool =
+  info "validating true"
+  true
+
+type RestrictiveValidator = ref object of EntryValidator
+
+type ApatheticSelector = ref object of EntrySelector
+method select(
+    self: ApatheticSelector, cand: RecordVal, others: seq[RecordVal]
+): RecordVal =
+  return cand
+
+method validate(self: RestrictiveValidator, cand: EntryCandidate): bool =
+  info "validating false"
+  false
+
+suite "KadDHT - PutVal":
```

Comment:
this passes 100%* due to replication factor of 2, meaning that all nodes will be getting the entry. with a rep factor of 1, the test is flakey. Can't say for sure, but current working hypothesis is that peer ids are rng'd, and selection of putval targets is based on peer id in this test. this selection doesn't always make the choice we expect.

Subject te review feedback, make GH issue to follow up before merging. 

This probably relates to the k-buckets divergence from spec that was rased in initial kad-dht PR